### ISSUE TITLE
slice appends sometimes have some of their content zeroed out in between tx calls

### ISSUE THREAD

**Issue by thehowl** [gnolang/gno#1167]
Reproducible test: https://github.com/gnolang/gno/commit/0e44fca3c20108500cf375967b0af08f5d909aa1

Without the workaround, test output:

```console
$ cd gno.land
$ go test -v -p 1 -run '/issue_1167' -timeout=30m ./cmd/gnoland
=== RUN   TestTestdata
=== RUN   TestTestdata/issue-1167
=== PAUSE TestTestdata/issue-1167
=== CONT  TestTestdata/issue-1167
    testscript.go:534: WORK=$WORK
        PATH=[...]
        GOTRACEBACK=system
        HOME=/no-home
        TMPDIR=$WORK/.tmp
        devnull=/dev/null
        /=/
        :=:
        $=$
        GOCOVERDIR=
        exe=
        USER_SEED_test1=source bonus chronic canvas draft south burst lottery vacant surface solve popular case indicate oppose farm nothing bullet exhibit title speed wink action roast
        USER_ADDR_test1=g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5
        GNOROOT=/home/howl/oc/gno
        GNOHOME=/home/howl/throw/tmp/TestTestdata2981033143/001/gno

        # Reproducible Test for https://github.com/gnolang/gno/issues/1167 (3.436s)
        > gnoland start
        starting logger: "/home/howl/throw/tmp/tmp.9A1tFRif12/gnoland-c719f008.log"
        [stdout]
        node started successfully

        # add contract (0.425s)
        > gnokey maketx addpkg -pkgdir $WORK -pkgpath gno.land/r/demo/xx -gas-fee 1000000ugnot -gas-wanted 2000000 -broadcast -chainid=tendermint_test test1
        [stdout]

        OK!
        GAS WANTED: 2000000
        GAS USED:   416224

        [stderr]
        Enter password.

        > stdout OK!
        # execute New (0.435s)
        > gnokey maketx call -pkgpath gno.land/r/demo/xx -func New -args X -gas-fee 1000000ugnot -gas-wanted 2000000 -broadcast -chainid=tendermint_test test1
        [stdout]
        ("" string)
        OK!
        GAS WANTED: 2000000
        GAS USED:   260345

        [stderr]
        Enter password.

        > stdout OK!
        # execute Delta for the first time (0.511s)
        > gnokey maketx call -pkgpath gno.land/r/demo/xx -func Delta -args X -gas-fee 1000000ugnot -gas-wanted 2000000 -broadcast -chainid=tendermint_test test1
        [stdout]
        ("1,1,1;" string)
        OK!
        GAS WANTED: 2000000
        GAS USED:   307608

        [stderr]
        Enter password.

        > stdout OK!
        > stdout '"1,1,1;" string'
        # execute Delta for the second time (0.432s)
        > gnokey maketx call -pkgpath gno.land/r/demo/xx -func Delta -args X -gas-fee 1000000ugnot -gas-wanted 2000000 -broadcast -chainid=tendermint_test test1
        [stdout]
        ("0,0,0;2,2,2;" string)
        OK!
        GAS WANTED: 2000000
        GAS USED:   309514

        [stderr]
        Enter password.

        > stdout OK!
        > stdout '1;2,2,2;" string'
        FAIL: testdata/issue-1167.txtar:21: no match for `1;2,2,2;" string` found in stdout

--- FAIL: TestTestdata (0.00s)
    --- FAIL: TestTestdata/issue-1167 (5.61s)

```

This was originally found in GnoChess: one of the head-banging-against-the-wall issues we had with GnoChess was that the realm was returning `a1a1` as the first move being made (`a1a1` is a representation of a Move struct, where if all of its three values are 0 then the String() method will return `a1a1`). This could not be replicated using Gno tests, where the behaviour didn't happen, but only with the live chain.

Introducing this [change](https://github.com/gnolang/gnochess/commit/7da3a8d3a656039f64ab11f524f32e08e438275a#diff-f7444df1c9cb12d6dbdd27500af70893da4507d232ec5d92b1cb1aa4ed114a73R429) fixed the problem. The underlying idea is that of re-creating and reallocating the moves slice, instead of copying it from the previous instance of `Position`.

**Comment by clockworkgr** [comment]
Possibly related: https://github.com/gnolang/gno/pull/1170

something-something-slice-allocation

**Comment by thehowl** [comment]
cc/ @deelawn

**Comment by deelawn** [comment]
I've done an analysis on this and was able to identify the root cause.  I'll start from the call to `clone()` and work my way up the stack.

### Inside `Position.clone()`:
`mv := p.Moves` -- The rhs already exists as a realm object.  The lhs is local and is NOT marked for realm persistence.  The value is used to compose a new instance of `Position` that is returned.

### Inside `Position.update()`:
`p := oldp.clone()` -- The rhs has just been initialized so the assignment to the new variable `p` results in another object NOT marked for realm persistence, even though the embedded and persisted `Moves` object exists within it.

`p.Moves = append(p.Moves, Move{counter, counter, counter})` -- A new `Move` instance is appended to the existing and persistent underlying array value.  It is within the defined capacity so no resizing occurs.  The lhs variable is not marked for persistence because the parent/base, `p`, is not yet a persistent object.

### Inside `Delta(string)`:
`n := g.Position.update()` -- a wash in regards to marking anything for persistence; `n` is not persistent

`g.Position = n` -- the lhs is persistent and being assigned to a newly created instance of `Position`, so it is marked as `newReal` in order to be persisted

### And then...
When it's time to finalize the transaction and handle CRUD operations for realm persistence objects, the realm identifies there is a newly created instance of `Position`. It creates and and proceeds to recurse over its children.  _BUT_ the only child to process is the existing `[]Move` instance.  Now examine this block of code: https://github.com/gnolang/gno/blob/master/gnovm/pkg/gnolang/realm.go#L407-L433

Note that `child` is the existing `[]Move` instance here.  It already has an existing reference so after incrementing it again, `rc` becomes 2 and the necessary block to persist the new `Move` object is skipped.  Even if the reference count were correct, this instance of `[]Move` is currently a real object, so it would be marked as dirty and the newly value would not be persisted.

So there are two things that need to be addressed here:
1. How to handle decrementing a reference to an object when it is reassigned and its parent object is overwritten
2. How to determine if a real object may be composed of new real objects that could be two or more layers deep due to transitive assignments

I'll dig into this more and come up with one or more proposals for how to fix these issues.

**Comment by jaekwon** [comment]
looking at this now.

**Comment by deelawn** [comment]
Fixed by #1305
